grpc/creds/creds.go

```diff
@@ -51,9 +55,73 @@ func (tc *transportCredentials) ClientHandshake(ctx context.Context, addr string
 	}
 }
 
+func (tc *transportCredentials) peerIsWhitelisted(peerState tls.ConnectionState) (bool, error) {
+	// If there's no whitelist, all clients are OK
+	if tc.whitelist == nil {
+		return true, nil
+	}
+
+	// Otherwise its time to start inspecting the peer's `VerifiedChains`
+	chains := peerState.VerifiedChains
+	if len(chains) < 1 {
+		return false, fmt.Errorf("peer had zero VerifiedChains")
+	}
+
+	/*
+	 * For each of the peer's verified chains we can look at the chain's leaf
+	 * certificate and check whether the subject common name is in the whitelist.
+	 * At least one chain must have a leaf certificate with a subject CN that
+	 * matches the whitelist
+	 *
+	 * Its important we process `VerifiedChains` instead of processing
+	 * `PeerCertificates` to ensure that we match the subject CN of the
+	 * leaf certificate that was verified in `conn.Handshake()`. To do otherwise
+	 * would allow an attacker to include a whitelisted certificate in
+	 * `PeerCertificates` that matched the whitelist but wasn't used in the chain
+	 * the server validated.
+	 */
+	var whitelisted bool
+	for _, chain := range chains {
+		leafSubjectCN := chain[0].Subject.CommonName
+		if _, ok := tc.whitelist[leafSubjectCN]; ok {
+			whitelisted = true
+		}
+	}
+
+	// If none of the peer's validated chains had a leaf certificate with a
+	// whitelisted CN then we have to reject the connection
+	if !whitelisted {
+		return false, fmt.Errorf(
+			"peer's verified TLS chains did not include a leaf certificate " +
+				"with a whitelisted subject CN")
+	}
+
+	// Otherwise, the peer is whitelisted! Come on in!
+	return true, nil
+}
+
 // ServerHandshake performs the TLS handshake for a server <- client connection
 func (tc *transportCredentials) ServerHandshake(rawConn net.Conn) (net.Conn, credentials.AuthInfo, error) {
-	return nil, nil, fmt.Errorf("boulder/grpc/creds: Server-side handshakes are not implemented")
+	// If there's no `serverConfig` this is a client-only `transportCredentials`
+	// and we should return an error
+	if tc.serverConfig == nil {
+		return nil, nil, fmt.Errorf(
+			"boulder/grpc/creds: Server-side handshake not supported without non-nil " +
+				"`serverConfig`")
+	}
+
+	// Perform the server <- client TLS handshake
+	conn := tls.Server(rawConn, tc.serverConfig)
+	if err := conn.Handshake(); err != nil {
+		return nil, nil, err
+	}
+
+	// If the peer isn't whitelisted, abort and return an error
+	if whitelisted, err := tc.peerIsWhitelisted(conn.ConnectionState()); !whitelisted {
+		return nil, nil, fmt.Errorf("boulder/grpc/creds: %s", err.Error())
```

Comment:
I think this can only return two set errors, instead of pre-prending the package name here could you just add it to the errors and directly return the `err`.